gateway-ha/src/main/java/io/trino/gateway/ha/handler/QueryIdCachingProxyHandler.java

```diff
@@ -118,6 +119,12 @@ else if (path.startsWith(TRINO_UI_PATH)) {
                 queryId = matcher.group(1);
             }
         }
+        if (!isNullOrEmpty(queryParams)) {
+            Matcher matcher = QUERY_ID_PARAM_PATTERN.matcher(queryParams);
+            if (matcher.matches()) {
+                queryId = matcher.group(1);
+            }
+        }
```

Comment:
Assuming that queryparam comes after `/ui` only, 
this code can be simplified with changing line 117 (`Matcher matcher = QUERY_ID_PATTERN.matcher(path);`) with below. No need for new regex pattern, which is very similiar to `QUERY_ID_PATTERN`.
I've tested using QUERY_ID_PATTERN matcher on your test cases and it passes fine
```java
Matcher matcher = QUERY_ID_PATTERN.matcher(
                    path + (isNullOrEmpty(queryParams) ? "" : "?" + queryParams));

Reply:
added a test case for redirects to illustrate a case where the queryid may not follow `/ui`. I will narrow the scope of the path matcher since it includes some patterns that will not appear in a path.